barretenberg/cpp/pil/avm/gadgets/range_check.pil

```diff
@@ -0,0 +1,188 @@
+include "../main.pil";
+include "../fixed/powers.pil";
+
+namespace range_check(256);
+    pol commit clk;
+
+    // Range check selector
+    pol commit sel_rng_chk;
+    sel_rng_chk * (1 - sel_rng_chk) = 0;
+
+    // Witnesses 
+    // Value to range check
+    pol commit value;
+    // Number of bits to check against
+    pol commit rng_chk_bits;
+
+    // Bit Size Columns
+    // It is enforced (further down) that the selected column is the lowest multiple of 16 that is greater than rng_chk_bits 
+    // e.g., rng_chk_bits = 10 ===> is_lte_u16, rng_chk_bits = 100 ==> is_lte_u112
+    // TODO: Maybe we can get rid of is_lte_u128 since it's implicit if we have sel_rng_chk and no other is_lte_x
+    pol commit is_lte_u16;
+    pol commit is_lte_u32;
+    pol commit is_lte_u48;
+    pol commit is_lte_u64;
+    pol commit is_lte_u80;
+    pol commit is_lte_u96;
+    pol commit is_lte_u112;
+    pol commit is_lte_u128;
+    is_lte_u16 * (1 - is_lte_u16) = 0;
+    is_lte_u32 * (1 - is_lte_u32) = 0;
+    is_lte_u48 * (1 - is_lte_u48) = 0;
+    is_lte_u64 * (1 - is_lte_u64) = 0;
+    is_lte_u80 * (1 - is_lte_u80) = 0;
+    is_lte_u96 * (1 - is_lte_u96) = 0;
+    is_lte_u112 * (1 - is_lte_u112) = 0;
+    is_lte_u128 * (1 - is_lte_u128) = 0;
+
+    // Mutual Exclusivity condition
+    is_lte_u16 + is_lte_u32 + is_lte_u48 + is_lte_u64 + is_lte_u80 + is_lte_u96 + is_lte_u112 + is_lte_u128 = sel_rng_chk;
+
+    // Eight 16-bit slice registers
+    pol commit u16_r0;
+    pol commit u16_r1;
+    pol commit u16_r2;
+    pol commit u16_r3;
+    pol commit u16_r4;
+    pol commit u16_r5;
+    pol commit u16_r6;
+    // This register has a (more expensive) set of constraint that enables dynamic range check of bit values between 0 and 16 bits
+    pol commit u16_r7;
+
+    // In each of these relations, the u16_r7 register contains the most significant 16 bits of value.
```

Comment:
@IlyasRidhuan "the most significant 16 bits of value." You meant after zero padding on the higher bits to express value with a number of bits which is a multiple of 16.